go/consensus/tendermint/abci/state.go

```diff
@@ -331,71 +231,81 @@ func (s *applicationState) afterHaltEpoch(ctx *Context) bool {
 	return currentEpoch > s.haltEpochHeight
 }
 
+func (s *applicationState) doInitChain(now time.Time) error {
+	s.blockLock.Lock()
+	defer s.blockLock.Unlock()
+
+	return s.doCommitOrInitChainLocked(now)
+}
+
 func (s *applicationState) doCommit(now time.Time) error {
-	// Save the new version of the persistent tree.
-	blockHash, blockHeight, err := s.deliverTxTree.SaveVersion()
-	if err == nil {
-		s.blockLock.Lock()
-		s.blockHash = blockHash
-		s.blockHeight = blockHeight
-		s.blockTime = now
-		s.blockLock.Unlock()
-
-		// Reset CheckTx state to latest version. This is safe because
-		// Tendermint holds a lock on the mempool for commit.
-		//
-		// WARNING: deliverTxTree and checkTxTree do not share internal
-		// state beyond the backing database.  The `LoadVersion`
-		// implementation MUST be written in a way to avoid relying on
-		// cached metadata.
-		//
-		// This makes the upstream `LazyLoadVersion` and `LoadVersion`
-		// unsuitable for our use case.
-		_, cerr := s.checkTxTree.LoadVersion(blockHeight)
-		if cerr != nil {
-			panic(cerr)
-		}
+	s.blockLock.Lock()
+	defer s.blockLock.Unlock()
+
+	_, stateRootHash, err := s.deliverTxTree.Commit(s.ctx, s.stateRoot.Namespace, s.stateRoot.Round+1)
```

Comment:
What's this new round thing that we use instead of height?

Reply:
It's the same thing, the only problem is that our MKVS has this concept of rounds (they are actually versions) because it was initially used only for runtime storage, which has a concept of rounds from the roothash module.

Now that this is more generally used it would probably make sense to change all "rounds" in MKVS to "versions".

Reply:
I've filed #2734 to make this more general now that we use MKVS in different places.